src/pages/study-page/study-results-page/study-results-page.tsx

```diff
@@ -92,4 +118,49 @@ export const StudyResultsPage = ({
       </>
     );
   }
+
+  async function handleFinishClick() {
+    if (isUpdating) {
+      return;
+    }
+
+    // update card scores
+    setIsUpdating(true);
+    await _updateAllCardScores();
+    setIsUpdating(false);
+
+    // navigation is handled with a side effect on `areResultsApplied`
+    setAreResultsApplied(true);
+  }
+
+  // TODO: it might be beneficial to have an endpoint to batch these (an array of {id, score})
+  // we are currently making requests equal to the number of seen lesson cards (marked correct/incorrect)
+  async function _updateAllCardScores() {
+    if (isUpdating) {
+      return;
+    }
+
+    const promises = [];
+
+    // create promise to update cards with new scores
+    for (const lessonCard of lessonCards) {
+      if (lessonCard.outcome === 'unseen') {
+        continue;
+      }
+
+      // spaced repetition: if correct, score is increased by 1; otherwise reset to 0
+      let newScore = 0;
+      if (lessonCard.outcome === 'correct') {
+        newScore = Math.min(lessonCard.score + 1, MAX_SCORE);
+      }
+
+      if (lessonCard.id) {
+        promises.push(cardsClient.updateCardScoreById(lessonCard.id, newScore));
+      } else {
+        throw new Error(`Failed to update lesson card: id was undefined`); // this shouldn't be possible...
```

Comment:
I am starting to think we should have two card types: one with the id required and the other with it optional. Then we would just use the optional type in the deck editor, so we can avoid all these checks.

Reply:
probably not apart of this PR but we could make `id` required on the current cards and then the new type would be something like this:

```ts
type NewCard = Omit<Card, 'id'> & Partial<Pick<Card, 'id'>>`
```
which will create a `Card` type with the `id` optional.